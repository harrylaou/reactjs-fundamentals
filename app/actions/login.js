export const LOGIN_USER_OK = 'LOGIN_USER_OK'
export const LOGOUT_USER = 'LOGOUT_USER'

export const loginUser = () => ({
  type: LOGIN_USER_OK
})

export const logoutUser = () => ({
  type: LOGOUT_USER
})