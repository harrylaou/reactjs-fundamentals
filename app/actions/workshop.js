export const RECEIVE_WORKSHOP = 'RECEIVE_WORKSHOP'


export const receiveWorkshop = workshop => ({
  type: RECEIVE_WORKSHOP,
  workshop
})