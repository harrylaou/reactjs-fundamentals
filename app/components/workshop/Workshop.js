import React, {Component} from 'react'
import {
  Card,
  CardActions,
  CardHeader,
  CardMedia,
  CardTitle,
  CardText
} from 'material-ui/Card'
import es6promise from 'es6-promise'
import 'isomorphic-fetch'

es6promise.polyfill()

const Workshop = props => {

  let workshop = props.workshop
  let id = props.id
  console.log("Workshop id", props.id)

  if (workshop && workshop.id != id) {
    props.fetchWorkshop(id)
  }
  let title = ''
  let price = ''
  let url = ''
  let avatar = ''

  if (workshop) {
    title = `${workshop.title}`
    price = `${workshop.price}`
    url = `${workshop.url}`
    avatar = `${workshop.instructors[0].avatar}`
  }

  return (
    <Card>
      <CardHeader title={title} subtitle={url} avatar={avatar}/>
      <CardActions/>
    </Card>
  )
}

export default Workshop
