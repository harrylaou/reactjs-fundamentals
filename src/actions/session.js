export const LOG_OUT = 'LOG_OUT'
export const LOGGED_IN = 'LOGGED_IN'

export const logout = () => ({
  type: LOG_OUT
})

export const loggedIn = token => ({
  type: LOGGED_IN,
  token
})
