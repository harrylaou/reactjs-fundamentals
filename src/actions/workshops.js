export const RECEIVE_WORKSHOPS = 'RECEIVE_WORKSHOPS'

export const receiveWorkshops = workshops => ({
  type: RECEIVE_WORKSHOPS,
  workshops
})
