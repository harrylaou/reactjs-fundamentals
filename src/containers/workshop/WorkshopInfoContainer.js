import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopInfo from '../../components/workshop/WorkshopInfo'
import * as api from '../../api'
import * as actions from '../../actions/workshop'

class WorkshopInfoContainer extends Component {
  constructor() {
    super()
    this.fetchWorkshop = this.fetchWorkshop.bind(this)
  }

  componentDidMount() {
    this.fetchWorkshop(this.props.params.workshop)
  }

  componentWillReceiveProps(nextProps) {
    // If you are thinking that performance could be improved by caching
    // the user, we agree. But we want to keep this example simple.
    // We are going to improve this code in the next lessons. We are building up :)
    if (this.props.params.workshop !== nextProps.params.workshop) {
      this.fetchWorkshop(nextProps.params.workshop)
    }
  }

  // Imagine we want to fetch another workshop from another component. How do you
  // think we could reuse this code?
  fetchWorkshop(workshop) {
    api
      .getWorkshop(workshop)
      .then((workshop) => {
        this.props.dispatch(actions.receiveWorkshop(workshop))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  render() {
    return (
      <WorkshopInfo
        {...this.props}
        workshop={this.props.workshop}
      />
    )
  }
}

WorkshopInfoContainer.propTypes = {
  params: React.PropTypes.object.isRequired,
  dispatch: React.PropTypes.func.isRequired,
  workshop: React.PropTypes.object
}

const mapStateToProps = (state) => ({
  workshop: state.workshop
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WorkshopInfoContainer)
