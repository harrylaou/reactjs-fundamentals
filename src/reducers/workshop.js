import { RECEIVE_WORKSHOP } from '../actions/workshop'

const workshop = (state = null, action) => {
  switch (action.type) {
    case RECEIVE_WORKSHOP: {
      return action.workshop
    }
    default:
      return state
  }
}

export default workshop
