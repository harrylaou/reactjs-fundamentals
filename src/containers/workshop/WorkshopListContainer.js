import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopList from '../../components/workshop/WorkshopList'
import * as api from '../../api'
import * as actions from '../../actions/workshops'
import withWidth from '../../utils/WithWidth'
import { withRouter } from 'react-router'

class WorkshopListContainer extends Component {
  constructor() {
    super()
    this.showWorkshop = this.showWorkshop.bind(this)
  }

  componentDidMount() {
    // Imagine we want to fetch workshops from another component. How do you
    // think we could reuse this code?
    api
      .getWorkshops()
      .then((workshops) => {
        this.props.dispatch(actions.receiveWorkshops(workshops))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  showWorkshop(workshop) {
    this.props.router.push(`/workshops/${workshop.id}`)
  }

  render() {
    return (
      <WorkshopList
        {...this.props}
        workshop={this.props.params.workshop}
        workshops={this.props.workshops}
        showWorkshop={this.showWorkshop}
        width={this.props.width}
      />
    )
  }
}

WorkshopListContainer.propTypes = {
  width: React.PropTypes.number.isRequired,
  router: React.PropTypes.object.isRequired,
  workshops: React.PropTypes.array,
  dispatch: React.PropTypes.func.isRequired,
  params: React.PropTypes.object.isRequired
}

const mapStateToProps = (state) => ({
  workshops: state.workshops
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(withWidth(WorkshopListContainer)))
