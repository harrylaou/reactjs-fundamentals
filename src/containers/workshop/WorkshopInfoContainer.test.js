import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import ConnectedWorkshopInfoContainer, {WorkshopInfoContainer} from './WorkshopInfoContainer'
import { mockFetchContainer } from '../../testUtils'
import getMuiTheme from 'material-ui/styles/getMuiTheme'

import FakeConfigureStore from 'redux-mock-store'
import ActualConfigureStore from '../../store'
import {Provider} from 'react-redux'
import * as actions from '../../actions/workshop'

import injectTapEventPlugin from 'react-tap-event-plugin'
injectTapEventPlugin()

describe('<UserProfileContainer />', () => {
  it('should render with default props', () => {
    const props = {
      params: { workshop: '123' }
    }

    const wrapper = shallow(
      <WorkshopInfoContainer {...props} />
    )

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })


  const workshopData = {
    workshop: {
      title: 'reactjs_fundamentals',
      price: 990,
      url: 'https://reactjs.academy/react-redux-training-london',
      instructors: [{
        name: 'Manolo',
        url: '',
        avatar: ''
      }]
    }
  }

  const initialState = {
    workshop: {
      title: 'reactjs_advanced',
      price: 990,
      url: 'https://reactjs.academy/react-redux-training-london',
      instructors: [{
        name: 'Paco',
        url: '',
        avatar: ''
      }]
    }
  }
  const mockStore = FakeConfigureStore()
  const props = {
    params: { workshop: '123' }
  }
  let store

  describe('<ConnectedWorkshopInfoContainer />', () => {
    it('should render connected component', () => {
      store = mockStore(initialState)
      const container = shallow(
        <ConnectedWorkshopInfoContainer
          {...props}
          store={store}
        />
      )

      expect(shallowToJson(container)).toMatchSnapshot()
    })
  })

  describe('Integration test: Mount + Wrapping Provider', () => {
    let wrapper

    beforeEach(()=>{
      store = mockStore(initialState)

      mockFetchContainer(workshopData.wprkshop)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedWorkshopInfoContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedWorkshopInfoContainer).length).toEqual(1)
    })

    it('should Prop matches with initialState', () => {
      expect(wrapper.find(WorkshopInfoContainer).prop('workshop')).toEqual(initialState.workshop)
    })

    it('should check action dispatch', () => {
      store.dispatch(actions.receiveWorkshop(workshopData.workshop))
      const action = store.getActions()

      expect(action[0].type).toBe(actions.RECEIVE_WORKSHOP)
    })
  })

  describe('Integration test: Mount + Wrapping Provider + Actual Store + Reducers', () => {
    let wrapper

    beforeEach(()=>{
      store = ActualConfigureStore(initialState)

      mockFetchContainer(workshopData.workshop)

      wrapper = mount(
        <Provider store={store} >
          <ConnectedWorkshopInfoContainer
            {...props}
          />
        </Provider>,
        {
          context: {
            muiTheme: getMuiTheme()
          },
          childContextTypes:{
            muiTheme: React.PropTypes.object.isRequired,
          }
        }
      )
    })

    it('should mount and map connect connected component', () => {
      expect(wrapper.find(ConnectedWorkshopInfoContainer).length).toEqual(1)
    })

    it('should check action dispatch', () => {
      //OR store.dispatch(actions.receiveWorkshop(workshopData))
      wrapper.find(WorkshopInfoContainer).props().dispatch(actions.receiveWorkshop(workshopData.workshop))

      expect(wrapper.find(WorkshopInfoContainer).prop('workshop')).toEqual(workshopData.workshop)
    })
  })
})