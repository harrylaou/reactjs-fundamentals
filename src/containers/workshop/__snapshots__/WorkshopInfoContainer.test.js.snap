// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfileContainer /> <ConnectedWorkshopInfoContainer /> should render connected component 1`] = `
<WorkshopInfoContainer
  dispatch={[Function]}
  params={
    Object {
      "workshop": "123",
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  workshop={
    Object {
      "instructors": Array [
        Object {
          "avatar": "",
          "name": "Paco",
          "url": "",
        },
      ],
      "price": 990,
      "title": "reactjs_advanced",
      "url": "https://reactjs.academy/react-redux-training-london",
    }
  }
/>
`;

exports[`<UserProfileContainer /> should render with default props 1`] = `
<WorkshopInfo
  params={
    Object {
      "workshop": "123",
    }
  }
/>
`;
